### PR TITLE
[api] Fix `Fields` in `CompositeTextMapPropagator`

### LINKED ISSUE #2328: CompositeTextMapPropagator.Fields are always empty
# Bug Report

List of [NuGet packages](https://www.nuget.org/profiles/OpenTelemetry) and version that you are using (e.g. `OpenTelemetry 1.0.2`):

```
OpenTelemetry 1.1.0
OpenTelemetry.Api 1.1.0
OpenTelemetry.Extensions.Hosting 1.0.0-rc7
```

Runtime version (e.g. `net461`, `net48`, `netcoreapp3.1`, `net5.0` etc. You can find this information from the `*.csproj` file):

 `net5.0` 

## Symptom

Using [CompositeTextMapPropagator](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/Propagation/CompositeTextMapPropagator.cs) results in empty Fields, so i'm unable to properly clear carrier object before reusing (as described in [Fields spec](https://github.com/open-telemetry/opentelemetry-specification/blob/7891a4216116d460591ef15437d0fe7d35309c2c/specification/context/api-propagators.md#fields)).

**What is the expected behavior?**

CompositeTextMapPropagator.Fields returns union of underlying propagators' fields

**What is the actual behavior?**

CompositeTextMapPropagator.Fields always returns empty set

## Reproduce

See https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Context/Propagation/CompositeTextMapPropagator.cs#L41

## Additional information

I faced this issue when implementing tracing on top of Confluent.Kafka. After error on consume, i'm sending message to the delayed retry topic, and after delay this message is produced to original topic again. I want to clear previously set tracing headers and properly set new ones (but preserve other headers from original message). For now, it's impossible, because i couldn't differ tracing headers from application-defined headers.

### PR DESCRIPTION
**Pull request by stevejgordon**
Fixes #2328

## Changes

- Updates CompositeTextMapPropagator.Fields to return a unioned set of fields based on all propagators it wraps. This set is calculated once during instance construction.
- Updates the ctor to discard null propagators in the `IEnumerable<TextMapPropagator>`.
- Removes use of mutable static `EmptyFields`.
- Updates tests to ensure expected fields and adds assertions to validate that each wrapped propagator is called for `Inject` and `Extract`.
- Relocates test classes under the correct folder structure (and updates their namespace) to match the `OpenTelemetry.Api` project.

## Reviewer notes:
I'm uncertain whether the `Fields` property should be considered immutable based on the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/context/api-propagators.md#fields). Most implementations (apart from `B3Propagator`) return a new `HashSet<string>` every time they are accessed, resulting in the set always containing the same elements. However, a custom implementation of `TextMapPropagator` could potentially modify the set each time `Fields` is accessed. I'm unsure if this needs to be considered. Currently, I've computed the union of fields once in the `CompositeTextMapPropagator` constructor to avoid recalculating it for each access of the `Fields` property. If necessary, we can transfer this logic to the getter to accommodate wrapper propagators dynamically changing their fields.

I noticed that the `B3Propagator`, untouched in the PR, currently uses a static `HashSet<string>` which `Fields` returns. This could be a bug since it is then possible for an instance to add items to that set. This is an edge case, but we could also fix it by returning a new `HashSet` each time so that instances can mutate a shared set.

### ~~Test Failures (locally)~~
~~I had issues with running the OpenTelemetry.Api.Tests with `net462` as it doesn't seem to use the latest code from the project. When debugging, I could see that it pulled in `OpenTelemetry.Api` from my package cache instead, but it did not include the ctor changes. I'd love some thoughts on why that's happening. I had a quick look but couldn't see anything obvious. I even updated the csproj to explicitly reference the project being tested. Is it just on my machine? Hopefully, the CI run will also show if this is an issue.~~

```
[xUnit.net 00:00:01.99]     OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_SingleTextMapPropagator [FAIL]
[xUnit.net 00:00:02.01]     OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_TestPropagator [FAIL]
[xUnit.net 00:00:02.04]     OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_UsingSameTag [FAIL]
  Failed OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_SingleTextMapPropagator [16 ms]
  Error Message:
   Assert.Equal() Failure: HashSets differ
Expected: ["custom-traceparent-1", "custom-tracestate-1"]
Actual:   []
  Stack Trace:
     at OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_SingleTextMapPropagator() in C:\Users\SteveGordon\Code\OTel\opentelemetry-dotnet\test\OpenTelemetry.Api.Tests\Trace\Propagation\CompositePropagatorTest.cs:line 58
  Failed OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_TestPropagator [11 ms]
  Error Message:
   Assert.Equal() Failure: Values differ
Expected: 4
Actual:   0
  Stack Trace:
     at OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_TestPropagator() in C:\Users\SteveGordon\Code\OTel\opentelemetry-dotnet\test\OpenTelemetry.Api.Tests\Trace\Propagation\CompositePropagatorTest.cs:line 68
  Failed OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_UsingSameTag [3 ms]
  Error Message:
   Assert.Equal() Failure: Values differ
Expected: 3
Actual:   0
  Stack Trace:
     at OpenTelemetry.Context.Propagation.Tests.CompositePropagatorTest.CompositePropagator_UsingSameTag() in C:\Users\SteveGordon\Code\OTel\opentelemetry-dotnet\test\OpenTelemetry.Api.Tests\Trace\Propagation\CompositePropagatorTest.cs:line 112
[xUnit.net 00:00:02.90]     OpenTelemetry.Trace.Tests.TracerTest.TracerConcurrencyTest [SKIP]
  Skipped OpenTelemetry.Trace.Tests.TracerTest.TracerConcurrencyTest [1 ms]

Failed!  - Failed:     3, Passed:   216, Skipped:     1, Total:   220, Duration: 2 s - OpenTelemetry.Api.Tests.dll (net462)
```

~~Debugging with 'Just my code' disabled, when I enter the ctor, it's coming from an external source:~~
image

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [x] Unit tests added/updated
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial changes
~* [ ] Changes in public API reviewed (if applicable)~

### CODE REVIEW

**Comment by stevejgordon** [comment]
It looks like the `net462` tests were fine in CI, so it's likely a local issue for me to figure out. The failures are unrelated and being addressed in #5744.

**Comment by Kielek** [comment]
@stevejgordon, I can confirm that it is on your machine, executed all tests locally from your branch. All are green (except couple ignored).

File: src/OpenTelemetry.Api/Context/Propagation/CompositeTextMapPropagator.cs [hunk 1]
```diff
@@ -11,8 +11,8 @@ namespace OpenTelemetry.Context.Propagation;
 /// </summary>
 public class CompositeTextMapPropagator : TextMapPropagator
 {
-    private static readonly ISet<string> EmptyFields = new HashSet<string>();
-    private readonly List<TextMapPropagator> propagators;
+    private readonly IReadOnlyList<TextMapPropagator> propagators;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Api/Context/Propagation/CompositeTextMapPropagator.cs:14]: Below in `Extract` and `Inject` we perform...

```csharp
foreach (var propagator in this.propagators)
{
   ...
}
```

IIRC switching this from `List<TextMapPropagator>` to `IReadOnlyList<TextMapPropagator>` will cause the `struct` `List<T>.Enumerator` to be boxed into `IEnumerator` which causes an allocation.

Let's go back to `List<TextMapPropagator>` or if you _really_ want to use `IReadOnlyList<TextMapPropagator>` you could switch the loops to do something like...

```csharp
for (int i = 0; i < this.propagtors.Count; i++)
{
   var propagator = this.propagators[i];
   ...
}
```

Comment (stevejgordon) [reply src/OpenTelemetry.Api/Context/Propagation/CompositeTextMapPropagator.cs:14]: Thanks for catching this @CodeBlanch. Indeed, it was allocating 40B. I have switched to `for` loops, as suggested.

**Review by CodeBlanch — APPROVED** [review]
LGTM
